### PR TITLE
Relax index signature checks for type any

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this change, a type without an index signature is considered related to a type with an index signature if the type of that index signature is `any`. For example:

```typescript
var x: { foo: number, [x: string]: any };
let y = { foo: 1, baz: 2 }; 
x = y;  // No longer an error
```

The assignment `x = y` was previously considered an error because `y` has no index signature. However, given that an index signature of type `any` doesn't actually constrain properties in the type, the assignment is now permitted.

This issue was originally reported by @joewood [here](https://github.com/Microsoft/TypeScript/pull/3823#issuecomment-125595400).

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5251,7 +5251,7 @@ namespace ts {
                     return indexTypesIdenticalTo(IndexKind.String, source, target);
                 }
                 let targetType = getIndexTypeOfType(target, IndexKind.String);
-                if (targetType) {
+                if (targetType && !(targetType.flags & TypeFlags.Any)) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Use `isTypeAny`

Comment (ahejlsberg) [inline src/compiler/checker.ts]: That would just add function call and a redundant null check.

**Comment by RyanCavanaugh** [comment]
:+1:

**Comment by ahejlsberg** [comment]
@JsonFreeman Agreed.

**Comment by JsonFreeman** [comment]
:+1:

**Pull request merged**